pennylane/ops/op_math/prod.py

```diff
@@ -247,30 +242,41 @@ def decomposition(self):
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
 
-        def mats_gen():
-            for ops in self.overlapping_ops:
-                if len(ops) == 1:
-                    yield (
-                        qml.matrix(ops[0])
-                        if isinstance(ops[0], qml.Hamiltonian)
-                        else ops[0].matrix()
-                    )
+        mats = []
+        for ops in self.overlapping_ops:
+            if len(ops) == 1:
+                batched = ops[0].batch_size is not None
+                mat = qml.matrix(ops[0]) if isinstance(ops[0], qml.Hamiltonian) else ops[0].matrix()
+                if self.batch_size and not batched:
+                    mats.append(qml.math.stack([mat for _ in range(self.batch_size)]))
                 else:
-                    mats_and_wires_gen = (
-                        (
-                            qml.matrix(op) if isinstance(op, qml.Hamiltonian) else op.matrix(),
-                            op.wires,
-                        )
-                        for op in ops
-                    )
+                    mats.append(mat)
 
-                    reduced_mat, _ = math.reduce_matrices(
-                        mats_and_wires_gen=mats_and_wires_gen, reduce_func=math.dot
+            else:
+                mats_and_wires_gen = (
+                    (
+                        qml.matrix(op) if isinstance(op, qml.Hamiltonian) else op.matrix(),
+                        op.wires,
                     )
+                    for op in ops
+                )
 
-                    yield reduced_mat
+                reduced_mat, _ = math.reduce_matrices(
+                    mats_and_wires_gen=mats_and_wires_gen, reduce_func=math.matmul
+                )
+                batched = any(op.batch_size is not None for op in ops)
```

Comment:
might some `op.batch_size` values ever be `0` rather than `None`, and if so would that break things?

Reply:
I believe they are never 0 nor 1, even though having `batch_size = 1` for a non-broadcasted operator would make more sense for me.

@dwierichs why was this decision taken?

Reply:
right, 1 is better than 0 😛 I won't block this PR with this concern, but if `Operator` has the default value of 1 for batch_size, I think the checking of batch_size throughout the codebase would be a lot simpler. This looks good to me, but I'd like to make that change after this PR (unless @dwierichs has concerns we're missing, that is)

Reply:
`batch_size=None` indicates a non-broadcasted operation, which to me is different from a broadcasted operation with dimension `1`. I see that one could argue that they are the same, because they contain the same amount of information. However, without any broadcasting, an additional axis of size `1` seems very unintuitive to me. Also, `qml.RX(np.array([0.4]), wires=0)` and `qml.RX(0.4, wires=0)` would yield the same batch dimension if we set `1` as the default that currently is `None`. 
Finally, all features that do not support a broadcasting axis yet would become unsupported even without broadcasting if we replace `None` by `1` and consider it as "dummy broadcasting"

Reply:
I was under the impression that broadcasting support is done on a case-by-case basis, essentially via `op.matrix()` or something like that. For this reason, I was presuming that an operator could simply ignore the concept of batching if the batch_size is 1.

> `qml.RX(np.array([0.4]), wires=0)` and `qml.RX(0.4, wires=0)` would yield the same batch dimension

Is this an implementation detail, or is there actual value gained from this distinction?

Reply:
Maybe this is just a subjective thing, but I actually like the distinction. From my perspective, passing a broadcasted axis of size `1` that is not _intended_ to be a broadcasting axis points to a user code error.

Reply:
But maybe this is something we should discuss in a broader context :)